samza-sql/src/main/java/org/apache/samza/sql/impl/SqlTableJoinUtils.java

```diff
@@ -0,0 +1,65 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.samza.operators.TableDescriptor;
+import org.apache.samza.serializers.JsonSerdeV2;
+import org.apache.samza.serializers.KVSerde;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.sql.data.SamzaSqlCompositeKey;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.storage.kv.RocksDbTableDescriptor;
+
+
+/**
+ * Helper class for Samza SQL table join with RocksDb local table implementation. This can be used
+ * by Samza SQL IO resolvers to resolve the IO config for an input table in the SQL statement.
+ */
+public class SqlTableJoinUtils {
```

Comment:
Looks like a cache of TableDescriptor objects - Are these objects expensive to create? why do we require this class?

Reply:
This is not about caching. So the issue is that this part of the code used to be in JoinTranslator directly, which means all Samza SQL table joins are assumed to be local table only. In reality, there is no reason we can't support join with remote tables. In samza SQL, IO-related information is all stored in  SqlIOConfig (used to be SqlSourceConfig) hence we decided to store TableDescriptor there too. SqlIOConfig is fully managed by SqlIOResolver, so it does not make sense to have JoinTranslator to create its own RocksDb table descriptor anymore. The downside with this is that all IOResolvers need to be able to create them for table joins. We currently have three IOResolvers in Samza SQL, Config-based, Test, and SamzaLi IOResolvers. So without duplicating the code in all three places, I created this helper class so that all three resolvers can reuse the code. 

For Samza Table, we should always return the same descriptor for the same tableId so there is a cache in this helper. 

Hope this explains. 

@vjagadish

Reply:
Peng, 
the only method in this class is 
```
new RocksDbTableDescriptor("JoinTable-" + sourceName)
+          .withSerde(KVSerde.of(keySerde, valueSerde));
```
If there are no caching requirements for table-descriptors, can we not invoke the constructor for `new RocksDbTableDescriptor` at the call-site?

Reply:
Let me chat with you offline on this.

Reply:
I just tried not caching the descriptor and it works so you're right that it's just one liner of constructor call without the caching. I've dropped the helper class and added the one liner in the IO resolvers directly. The descriptor is a light weight object so it is okay to not cache it.